### PR TITLE
Prevent BatchNorm stats from being changed on model initialization from `yaml`

### LINKED ISSUE #17469: why initialize running_var is  0.8999
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question
```
def check_bn_init(model):
    for name, m in model.named_modules():
        if isinstance(m, nn.BatchNorm2d):
            print(f"{name} running_var initial value: {m.running_var[0].item()}")
            break

scale = 'l'
from ultralytics import YOLO
model1 = YOLO(f'yolo11{scale}.yaml', verbose=False).model
check_bn_init(model1)


# model.0.bn running_var initial value: 0.8999999761581421
```
### Additional

_No response_

### PR DESCRIPTION
**Pull request by Y-T-G**
Closes https://github.com/ultralytics/ultralytics/issues/17469

If we don't put the model in `eval` mode on initialization, the `BatchNorm` stats get changed:

**W/O PR**
```python
from torch import nn
from ultralytics import YOLO
def check_bn_init(model):
    for name, m in model.named_modules():
        f = True
        if isinstance(m, nn.BatchNorm2d) and any( m.running_var != 1):
            print(f"{name} running_var initial value: {m.running_var[0].item()}")
            f = False
            break
    if f:
        print("All running_var set to 1")
model = YOLO("yolo11n.yaml")
check_bn_init(model)
```
`model.model.0.bn running_var initial value: 0.8999999761581421`

```python
In [4]: model.model.model[-1].cv2[0][0].bn.running_var
Out[4]: 
tensor([0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000,
        0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000, 0.9000])
```

BatchNorm stats should only be allowed to change during training.

**W/ PR**
```python
from torch import nn
from ultralytics import YOLO
def check_bn_init(model):
    for name, m in model.named_modules():
        f = True
        if isinstance(m, nn.BatchNorm2d) and any( m.running_var != 1):
            print(f"{name} running_var initial value: {m.running_var[0].item()}")
            f = False
            break
    if f:
        print("All running_var set to 1")
model = YOLO("yolo11n.yaml")
check_bn_init(model)
```
`All running_var set to 1`

```python
In [2]: model.model.model[-1].cv2[0][0].bn.running_var
Out[2]: tensor([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.])
```

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Improved model initialization to ensure accurate stride calculation without affecting training behavior. 🛠️

### 📊 Key Changes
- Temporarily sets the model to evaluation mode before calculating strides to prevent unwanted changes to batch statistics.
- Ensures the model is switched back to training mode immediately after stride calculation.
- Maintains correct training state and initialization for model components.

### 🎯 Purpose & Impact
- Prevents accidental updates to batch normalization statistics during stride calculation, leading to more stable and reliable training. 👍
- Ensures model components are properly initialized without side effects, improving consistency for users training custom models.
- Enhances robustness and reliability of the model setup process, especially for advanced tasks like segmentation and pose estimation.

### CODE REVIEW

**Comment by Laughing-q** [comment]
@Y-T-G Will this impact training?

**Comment by Y-T-G** [comment]
@Laughing-q From my testing on PASCAL VOC when training from scratch, the loss and mAP were the same.

**main**
```python
Starting training for 1 epochs...

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
        1/1       2.6G      3.299      4.737      3.821         31        640: 100%|██████████| 1
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██
                   all       4952      12032    0.00206      0.117    0.00185   0.000464

1 epochs completed in 0.051 hours.
Optimizer stripped from runs/detect/train24/weights/last.pt, 5.5MB
Optimizer stripped from runs/detect/train24/weights/best.pt, 5.5MB

Validating runs/detect/train24/weights/best.pt...
WARNING ⚠️ validating an untrained model YAML will result in 0 mAP.
Ultralytics 8.3.48 🚀 Python-3.10.14 torch-2.5.1+cu124 CUDA:0 (Tesla T4, 15934MiB)
YOLO11n summary (fused): 238 layers, 2,586,052 parameters, 0 gradients, 6.3 GFLOPs
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██
                   all       4952      12032    0.00206      0.117    0.00185   0.000465
             aeroplane        204        285   0.000839     0.0842   0.000464   8.23e-05
               bicycle        239        337    0.00263      0.228    0.00213   0.000573
                  bird        282        459    0.00174     0.0174   0.000954   0.000274
                  boat        172        263          0          0          0          0
                bottle        212        469          0          0          0          0
                   bus        174        213   0.000816     0.0141   0.000423   4.23e-05
                   car        721       1201    0.00366      0.225    0.00297     0.0007
                   cat        322        358    0.00349      0.483    0.00357   0.000829
                 chair        417        756    0.00156     0.0767   0.000877   0.000256
                   cow        127        244          0          0          0          0
           diningtable        190        206          0          0          0          0
                   dog        418        489    0.00586      0.166    0.00347    0.00101
                 horse        274        348    0.00426     0.0144    0.00218    0.00061
             motorbike        222        325    0.00344      0.258    0.00287   0.000669
                person       2007       4528    0.00291      0.668     0.0119    0.00265
           pottedplant        224        480          0          0          0          0
                 sheep         97        242    0.00463    0.00413    0.00233   0.000698
                  sofa        223        239      0.004     0.0418    0.00217   0.000775
                 train        259        282   0.000675     0.0638   0.000364   8.06e-05
             tvmonitor        229        308   0.000767    0.00325   0.000422   4.22e-05
Speed: 0.1ms preprocess, 1.3ms inference, 0.0ms loss, 0.8ms postprocess per image
```

**PR**
```python
Starting training for 1 epochs...

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
        1/1       2.6G      3.299      4.737      3.821         31
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██
                   all       4952      12032    0.00206      0.117    0.00185   0.000464

1 epochs completed in 0.051 hours.
Optimizer stripped from runs/detect/train23/weights/last.pt, 5.5MB
Optimizer stripped from runs/detect/train23/weights/best.pt, 5.5MB

Validating runs/detect/train23/weights/best.pt...
WARNING ⚠️ validating an untrained model YAML will result in 0 mAP.
Ultralytics 8.3.48 🚀 Python-3.10.14 torch-2.5.1+cu124 CUDA:0 (Tesla T4, 15934MiB)
YOLO11n summary (fused): 238 layers, 2,586,052 parameters, 0 gradients, 6.3 GFLOPs
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██
                   all       4952      12032    0.00206      0.117    0.00185   0.000465
             aeroplane        204        285   0.000839     0.0842   0.000464   8.23e-05
               bicycle        239        337    0.00263      0.228    0.00213   0.000573
                  bird        282        459    0.00174     0.0174   0.000954   0.000274
                  boat        172        263          0          0          0          0
                bottle        212        469          0          0          0          0
                   bus        174        213   0.000816     0.0141   0.000423   4.23e-05
                   car        721       1201    0.00366      0.225    0.00297     0.0007
                   cat        322        358    0.00349      0.483    0.00357   0.000829
                 chair        417        756    0.00156     0.0767   0.000877   0.000256
                   cow        127        244          0          0          0          0
           diningtable        190        206          0          0          0          0
                   dog        418        489    0.00586      0.166    0.00347    0.00101
                 horse        274        348    0.00426     0.0144    0.00218    0.00061
             motorbike        222        325    0.00344      0.258    0.00287   0.000669
                person       2007       4528    0.00291      0.668     0.0119    0.00265
           pottedplant        224        480          0          0          0          0
                 sheep         97        242    0.00463    0.00413    0.00233   0.000698
                  sofa        223        239      0.004     0.0418    0.00217   0.000775
                 train        259        282   0.000675     0.0638   0.000364   8.06e-05
             tvmonitor        229        308   0.000767    0.00325   0.000422   4.22e-05
Speed: 0.1ms preprocess, 1.3ms inference, 0.0ms loss, 0.8ms postprocess per image
```

**Comment by Laughing-q** [comment]
@Y-T-G This is started from a pt weight? if we are starting training from yaml file I guess it'll get slightly different results? since the initialized values in BN are different.

**Comment by Y-T-G** [comment]
This might be a bug:

`WARNING ⚠️ validating an untrained model YAML will result in 0 mAP.`

https://github.com/ultralytics/ultralytics/blob/6c65d756e4886828ac4060472348a49668a9fc3e/ultralytics/engine/validator.py#L123-L124

Maybe we add `and model is None:`.

**Comment by Y-T-G** [comment]
@Laughing-q I started from `yolo11n.yaml`. I think the BatchNorm difference is very small to affect the results noticeably. The training should be starting from different initialization after PR, but the effect is too small to make any difference.

**Comment by Laughing-q** [comment]
@Y-T-G I'm still curious about how it gonna affect the final results of the training. I launched two 100 epochs yolo11n training on COCO with/without this PR to see how it goes. https://wandb.ai/laughing/BN-initialization-test

**Comment by Laughing-q** [comment]
@Y-T-G @glenn-jocher ok guys the experiments showing results by this PR is better. Perhaps we should merge this one. I've updated it a bit to keep updates under the `if isinstance(m, Detect)` condition and revert the mode back to `train` after we get the `stride`.
pic-241212-1325-34
@glenn-jocher FYI do you think is there another way to get `stride` variable without actual doing an inference?

**Comment by Laughing-q** [comment]
@Y-T-G I see, I think it's better to revert it to train mode since when you create a model from nn.Module it's train mode by default, better to remain it after the stride calculation.

**Comment by Laughing-q** [comment]
@Y-T-G @glenn-jocher Since recently I've been doing experiments based on yolo11s model, and I also launched a training with epochs=600 last week on yolo11s within this PR. Hence an update on the experiment results:
mAP on main: 46.3
mAP on this PR: 46.1
Probably we better hold this PR for a while?

**Comment by glenn-jocher** [comment]
@Y-T-G have you tried using the smart_inference_mode() context manager here instead of torch.no_grad()? It applies inference_mode() selectively based on the torch version if it permits it, which is a slightly better version of torch.no_grad(), but not sure if this will prevent BN changes.

**Comment by Laughing-q** [comment]
@Y-T-G @glenn-jocher Guys updates here:
- The updated BN stats in model initialization does not infect the training actually when training on large dataset like COCO:
   - And the reason why I got different results before, is because the `albumentations`, training is not deterministic when has `albumentations>=1.4.21` installed, I launched two more trainings last week with `albumentations==1.4.20` and got exactly the same results.
   - See: https://wandb.ai/laughing/BN-initialization-test?nw=nwuserlaughing, I got exactly the same results on every loss and every metric in my new tests with/without this PR, also you can check the git commit version there if you want, one on main and one within this PR changes.

I have a simple function to check and compare the BN stats between two models, and checked for the two models trained on COCO, the BN stats are exactly same.
```python
from torch import nn
from ultralytics import YOLO
def compare_bn(model1, model2):
    for m1, m2 in zip(list(model1.modules()), list(model2.modules())):
        # print(m1, m2)
        if isinstance(m1, nn.BatchNorm2d) and isinstance(m2, nn.BatchNorm2d):
            if (m1.running_var != m2.running_var).sum():
                print(f"different BN stats: {m1.running_var[0]} vs {m2.running_var[0]}")
compare_bn(
    YOLO("yolo11n-base.pt"),
    YOLO("yolo11n-PR.pt"),
)
```

I guess it's because there are a lot iterations and updated BN stats drowned the very first update. Hence I did run tests locally on small dataset like coco128:
- The metric and loss curves are very closely matched actually
pic-241230-1502-57
- checked the BN stats with the script, it's different
pic-241230-1516-14

Conclusion: this issue would infect training on a small number of iteration(usually a small dataset or smaller epochs) a bit, and no impact on large number of iterations(usually large dataset or more epochs). 
@glenn-jocher This seems tiny but based on the fact that we should not update the BN stats at all in model initialization, how about we merge the PR to make things right, the reason I wanted to hold this PR is because it seemed hurting the mAP in my past tests, but now I'm sure it's caused by the undeterminism of `albumentations` package. And @Y-T-G please let us know if you spot any related issues after merging.
@glenn-jocher As the alternative PR https://github.com/ultralytics/ultralytics/pull/18255, I checked and it's hard to find a decent value for even a 80/20 solution, since the updated BN stats are shifted differently and it's hard to find a value to mitigate all of them. 
For example:
here's some BN stats in some layers when fill the value with `0.447`:
```bash
model.model.0.bn running_var initial value: 0.9000130295753479
model.model.1.bn running_var initial value: 0.9097357392311096
model.model.2.cv1.bn running_var initial value: 0.9265685677528381
model.model.2.cv2.bn running_var initial value: 0.9280524849891663
model.model.2.m.0.cv1.bn running_var initial value: 0.9156420826911926
model.model.2.m.0.cv2.bn running_var initial value: 0.9271636009216309
model.model.3.bn running_var initial value: 0.9371340870857239
model.model.4.cv1.bn running_var initial value: 0.9062875509262085
model.model.4.cv2.bn running_var initial value: 0.9182401299476624
model.model.4.m.0.cv1.bn running_var initial value: 0.9125862717628479
model.model.4.m.0.cv2.bn running_var initial value: 0.9075859189033508
model.model.5.bn running_var initial value: 0.9050173163414001
model.model.6.cv1.bn running_var initial value: 0.9084212183952332
model.model.6.cv2.bn running_var initial value: 0.9147871732711792
model.model.6.m.0.cv1.bn running_var initial value: 0.9183041453361511
```
and if we bump up to `40`:
```bash
model.model.0.bn running_var initial value: 1.4613336324691772
model.model.1.bn running_var initial value: 0.9295873045921326
model.model.2.cv1.bn running_var initial value: 0.9201989769935608
model.model.2.cv2.bn running_var initial value: 0.9819196462631226
model.model.2.m.0.cv1.bn running_var initial value: 0.9020330309867859
model.model.2.m.0.cv2.bn running_var initial value: 0.9169237613677979
model.model.3.bn running_var initial value: 0.9019262790679932
model.model.4.cv1.bn running_var initial value: 0.9503502249717712
model.model.4.cv2.bn running_var initial value: 0.9259657859802246
model.model.4.m.0.cv1.bn running_var initial value: 0.9115145802497864
model.model.4.m.0.cv2.bn running_var initial value: 0.9085854887962341
model.model.5.bn running_var initial value: 0.9147798418998718
model.model.6.cv1.bn running_var initial value: 0.9054678678512573
model.model.6.cv2.bn running_var initial value: 0.9117942452430725
model.model.6.m.0.cv1.bn running_var initial value: 0.9185563325881958
```
the first stat exceeds the default `1.0` and others remain less than `1.0`.

**Comment by glenn-jocher** [comment]
@Laughing-q should we go ahead and merge this?

**Comment by Laughing-q** [comment]
@glenn-jocher yeah let's do that!
